### ISSUE TITLE
The training stops without triggering errors

### ISSUE THREAD

**Issue by ipietri** [SanghunYun/UDA_pytorch#15]
Hi,

I'm trying to run the code with the same dataset you published but the training stops without triggering any errors. I'm wondering if you have faced this problem before. 

See image below,
Thanks

![image](https://user-images.githubusercontent.com/70186411/136669614-2a6d0632-16c0-4e76-81ae-a192cd6825c9.png)

**Comment by kbwzy** [comment]
the same issue.
tensorflow-gpu                   2.6.0
torch                            1.9.0

**Comment by hahah181292** [comment]
I wonder  tha because the label dataset is too small, so the program can run very quickly without errors

**Comment by Liu-Jingyao** [comment]
The same issue on Colab. This is probably not because it was running too fast, but the fine tuning was interrupted for some unknown reason

**Comment by Liu-Jingyao** [comment]
This is because the author set that the training data should not be repeated used when non-uda mode, and I am asking the reason for doing so.
https://github.com/SanghunYun/UDA_pytorch/issues/18#issue-1067255803

**Comment by hahah181292** [comment]
> This is because the author set that the training data should not be repeated used when non-uda mode, and I am asking the reason for doing so. [#18 (comment)](https://github.com/SanghunYun/UDA_pytorch/issues/18#issue-1067255803)
Hello, have you know the reason for doing so ?

**Comment by Liu-Jingyao** [comment]
> > This is because the author set that the training data should not be repeated used when non-uda mode, and I am asking the reason for doing so. [#18 (comment)](https://github.com/SanghunYun/UDA_pytorch/issues/18#issue-1067255803)
> > Hello, have you know the reason for doing so ?

Hello. I've done a lot of research but I still don't know what this convoluted setting means. However, I observed that the model works well and played its due performance when I just made the training records repeatable in non-uda mode. 
It's hard to make sense in theory or practice. Maybe we should just ignore this and make our own version.

**Comment by hahah181292** [comment]
> > > This is because the author set that the training data should not be repeated used when non-uda mode, and I am asking the reason for doing so. [#18 (comment)](https://github.com/SanghunYun/UDA_pytorch/issues/18#issue-1067255803)
> > > Hello, have you know the reason for doing so ?
> 
> Hello. I've done a lot of research but I still don't know what this convoluted setting means. However, I observed that the model works well and played its due performance when I just made the training records repeatable in non-uda mode. It's hard to make sense in theory or practice. Maybe we should just ignore this and make our own version.

Thank you for your reply! Do you mean to set sup_iter to repeat_dataloader? But I found that train_eval mode could not be carried out under non-uda, because under the non-uda mode don't have eval_iter. How did you determine which step achieved the best effect? I am looking forward to your generous comments.

**Comment by Liu-Jingyao** [comment]
> > > > This is because the author set that the training data should not be repeated used when non-uda mode, and I am asking the reason for doing so. [#18 (comment)](https://github.com/SanghunYun/UDA_pytorch/issues/18#issue-1067255803)
> > > > Hello, have you know the reason for doing so ?
> > 
> > 
> > Hello. I've done a lot of research but I still don't know what this convoluted setting means. However, I observed that the model works well and played its due performance when I just made the training records repeatable in non-uda mode. It's hard to make sense in theory or practice. Maybe we should just ignore this and make our own version.
> 
> Thank you for your reply! Do you mean to set sup_iter to repeat_dataloader? But I found that train_eval mode could not be carried out under non-uda, because under the non-uda mode don't have eval_iter. How did you determine which step achieved the best effect? I am looking forward to your generous comments.

A late reply. You can add the eval_iter by refactoring the data structure used to store iterators, just like what you do in supervised learning.   For example, I changed the confusing data_iter array in main.py and train.py to a readable dict, just like: 
``` python3
# main.py
# ... ...
data = load_data(cfg)
data_iter = dict()
if cfg.uda_mode:
    # ... ...
    if cfg.mode == 'train':
        data_iter['sup_iter'] = data.sup_data_iter()
        data_iter['unsup_iter'] = data.unsup_data_iter()
    elif cfg.mode == 'train_eval':
        data_iter['sup_iter'] = data.sup_data_iter()
        data_iter['unsup_iter'] = data.unsup_data_iter()
        data_iter['eval_iter'] = data.eval_data_iter()
    else:
        data_iter['eval_iter'] = data.eval_data_iter()
else:
    if cfg.mode == 'train':
        data_iter['sup_iter'] = data.sup_data_iter()
    elif cfg.mode == 'train_eval':
        data_iter['sup_iter'] = data.sup_data_iter()
        data_iter['eval_iter'] = data.eval_data_iter()
    else:
        data_iter['eval_iter'] = data.eval_data_iter()
```

``` python3
# train.py
# ... ...
if 'sup_iter' in data_iter.keys():
    self.sup_iter = self.repeat_dataloader(data_iter['sup_iter'])
if 'unsup_iter' in data_iter.keys():
    self.unsup_iter = self.repeat_dataloader(data_iter['unsup_iter'])
if 'eval_iter' in data_iter.keys():
    self.eval_iter = data_iter['eval_iter']
```
You may have found your own solution, but I still wish my practice will inspire you a bit.  Feel free to ask me more.

**Comment by hahah181292** [comment]
> > > > > This is because the author set that the training data should not be repeated used when non-uda mode, and I am asking the reason for doing so. [#18 (comment)](https://github.com/SanghunYun/UDA_pytorch/issues/18#issue-1067255803)
> > > > > Hello, have you know the reason for doing so ?
> > > 
> > > 
> > > Hello. I've done a lot of research but I still don't know what this convoluted setting means. However, I observed that the model works well and played its due performance when I just made the training records repeatable in non-uda mode. It's hard to make sense in theory or practice. Maybe we should just ignore this and make our own version.
> > 
> > 
> > Thank you for your reply! Do you mean to set sup_iter to repeat_dataloader? But I found that train_eval mode could not be carried out under non-uda, because under the non-uda mode don't have eval_iter. How did you determine which step achieved the best effect? I am looking forward to your generous comments.
> 
> A late reply. You can add the eval_iter by refactoring the data structure used to store iterators, just like what you do in supervised learning. For example, I changed the confusing data_iter array in main.py and train.py to a readable dict, just like:
> 
> ```python
> # main.py
> # ... ...
> data = load_data(cfg)
> data_iter = dict()
> if cfg.uda_mode:
>     # ... ...
>     if cfg.mode == 'train':
>         data_iter['sup_iter'] = data.sup_data_iter()
>         data_iter['unsup_iter'] = data.unsup_data_iter()
>     elif cfg.mode == 'train_eval':
>         data_iter['sup_iter'] = data.sup_data_iter()
>         data_iter['unsup_iter'] = data.unsup_data_iter()
>         data_iter['eval_iter'] = data.eval_data_iter()
>     else:
>         data_iter['eval_iter'] = data.eval_data_iter()
> else:
>     if cfg.mode == 'train':
>         data_iter['sup_iter'] = data.sup_data_iter()
>     elif cfg.mode == 'train_eval':
>         data_iter['sup_iter'] = data.sup_data_iter()
>         data_iter['eval_iter'] = data.eval_data_iter()
>     else:
>         data_iter['eval_iter'] = data.eval_data_iter()
> ```
> 
> ```python
> # train.py
> # ... ...
> if 'sup_iter' in data_iter.keys():
>     self.sup_iter = self.repeat_dataloader(data_iter['sup_iter'])
> if 'unsup_iter' in data_iter.keys():
>     self.unsup_iter = self.repeat_dataloader(data_iter['unsup_iter'])
> if 'eval_iter' in data_iter.keys():
>     self.eval_iter = data_iter['eval_iter']
> ```
> 
> You may have found your own solution, but I still wish my practice will inspire you a bit. Feel free to ask me more.

Thank you very much for your reply! Your method solved the problem that train_eval mode don't have eval_iter  during the training, which helped me a lot. However, I met the following problems during the verification. I think the parameters in the eval.json file were not uploaded successfully (but the parameters in the non-uda.json file were), so I don't know what went wrong now. I am very sorry to bother you again, and I hope to get your guidance!
![{}F%5LX C 2TAO$IH@`YI1I](https://user-images.githubusercontent.com/94736343/162097687-b5110afd-fbcf-4994-b803-e8f1d0dac84e.png)

**Comment by LISCARqaq** [comment]
> Hi,
> 
> I'm trying to run the code with the same dataset you published but the training stops without triggering any errors. I'm wondering if you have faced this problem before.
> 
> See image below, Thanks
> 
> ![image](https://user-images.githubusercontent.com/70186411/136669614-2a6d0632-16c0-4e76-81ae-a192cd6825c9.png)

Hello, I have the same issue too. Have you fixed the problem and how to do?

**Comment by Liu-Jingyao** [comment]
> > Hi,
> > I'm trying to run the code with the same dataset you published but the training stops without triggering any errors. I'm wondering if you have faced this problem before.
> > See image below, Thanks
> > ![image](https://user-images.githubusercontent.com/70186411/136669614-2a6d0632-16c0-4e76-81ae-a192cd6825c9.png)
> 
> Hello, I have the same issue too. Have you fixed the problem and how to do?

Hello. Yes, you can find my solution in my previous comments.
